---
title: "Application example"
author: "Noriaki Sato"
package: wcGeneSummary
output: 
    BiocStyle::html_document:
        toc_float: true
vignette: >
    %\VignetteIndexEntry{wcGeneSummary}
    %\VignetteEngine{knitr::knitr}
    %\VignetteEncoding{UTF-8}  
---

# Application examples

We demonstrate an use case of the package, which investigates transcriptomic changes induced by BK polyomavirus (BKPyV) infection in renal proximal tubular epithelial cells ([Assetta et al. 2016](https://pubmed.ncbi.nlm.nih.gov/27381292/)). Differentially expressed mRNAs in 3 days post-infection were obtained, and down-regulated mRNAs in BKPyV infected cells were examined.

## Load the necessary packages and genes

```{r load2, warning=FALSE, message=FALSE, cache=FALSE}
library(wcGeneSummary)
library(org.Hs.eg.db)
library(ggplot2)
library(ReactomePA);library(clusterProfiler)
library(ggraph);library(igraph)
library(ggforce) ## For genePathPlot
degs <- wcGeneSummary:::d3degDownAssetta2016
length(degs)
degs
```

## Enrichment analysis

First, we perform enrichment analysis using ReactomePA.
From the enrichment analysis results, the cluster is related to transcriptional regulation by TP53.

```{r ea, warning=FALSE, message=FALSE, cache=TRUE}
## Convert to ENTREZID
entre <- AnnotationDbi::select(org.Hs.eg.db, keytype="SYMBOL",
                               keys = degs, columns = "ENTREZID")$ENTREZID
pway <- setReadable(enrichPathway(entre), org.Hs.eg.db)
sigpway <- subset(pway@result, p.adjust<0.05)
sigpway$Description
cnetplot(pway)

## Genes involved in significant pathways
excheck <- unlist(unique(sapply(sigpway$geneID,
                                function (x) strsplit(x,"/"))))
excheck
```

We store the name of enriched pathways in the network for the downstream analysis.

```{r netenrich, cache=FALSE}
netreac <- wcGeneSummary(degs,
                         enrich="reactome",
                         plotType="network",
                         numWords=50,
                         colorText=TRUE)
```

## Text mining the gene summaries
Next we perform the plain function producing a correlation network, with showing the top-genes related to high-frequency words in the text in RefSeq summary. We obtained the list of these genes from geneCount slot.

```{r basic2app, warning=FALSE, message=FALSE, cache=TRUE, fig.width=10, fig.height=10}
net1 <- wcGeneSummary(excheck,
                      plotType="network",
                      colorText=TRUE,
                      numWords=30,
                      corThresh=0.5,
                      genePlot=TRUE,
                      genePlotNum=5,
                      edgeLink=FALSE,
                      genePathPlot="reactome")
net1@net
net1@geneCount
top <- names(net1@geneCount[net1@geneCount>=5])
top
```

## Text mining the available literature

These genes are further queried for PubMed information.
First, show the network for the article titles.

```{r absttitleapp, fig.height=9, fig.width=9}
titlenet <- wcAbst(top,
                   sortOrder="relevance",
                   target="title",
                   plotType="network",
                   colorText=TRUE,
                   madeUpperGenes=FALSE,
                   corThresh=0.2,
                   preset=TRUE,
                   retMax=40,
                   numWords=40,
                   edgeLink=FALSE)
titlenet@net
```

Obtain and show the network for the article abstract.

```{r abstmainapp, fig.height=9, fig.width=9}
abstnet <- wcAbst(top,
                   target="abstract",
                   plotType="network",
                   colorText=TRUE,
                   madeUpperGenes=FALSE,
                   corThresh=0.2,
                   preset=TRUE,
                   retMax=40,
                   numWords=40,
                   edgeLink=FALSE)
abstnet@net
```

## Combine and compare networks

From the RefSeq summary and articles related to important genes, the cluster could have functionality of DNA damage response, which is also upregulated by BKPyV infection. These networks can be combined to find intersections and differences. We can see that in addition to Reactome pathway names, plenty of information could be obtained and summarized by querying other databases, which could aid in interpreting clusters of genes and hypothesis generation.

```{r combineapp, cache=F, fig.height=20, fig.width=20}
compareWordNet(list(abstnet, titlenet, netreac, net1),
               titles=c("Abstract","Title","Reactome","RefSeq"))
```

From the network, DNA damage repair pathway, especially nucleotide excision repair related to DDB2, METTL14,and RAD17 might be related to BKPyV infection, which cannot be prioritize based on log2FoldChange or enrichment analysis.

```{r ddr, cache=T}
conet <- compareWordNet(list(netreac,
                    net1,
                    titlenet,
                    abstnet),
               returnNet = TRUE)

ddrNms <- NULL
for (nm in names(V(conet))) {
    if (tolower(nm) %in% c("dna","damage","repair")) {
        ddrNms <- c(ddrNms, names(neighbors(conet, nm)))
    }
}
ddrNms

ddrRelated <- induced.subgraph(conet,
                 names(V(conet)) %in% unique(ddrNms))
V(ddrRelated)$SYMBOL <- names(V(ddrRelated)) %in% keys(org.Hs.eg.db,"SYMBOL")
ggraph(ddrRelated)+
    geom_edge_diagonal2(color="grey80")+
    geom_node_point(aes(color=SYMBOL))+
    geom_node_text(aes(label=name, color=SYMBOL),check_overlap=TRUE, repel=TRUE,
                   bg.color = "white", segment.color="black",
                   bg.r = .15)+
    scale_color_manual(values=c("steelblue","tomato"))+
    theme_graph()
```


The network can be obtained by `returnNet=TRUE`, which can be used for downstream analysis like assessment of degrees and community detection.

```{r combineNet, cache=F, fig.height=10, fig.width=10}

conetDeg <- igraph::degree(conet)
conetDeg[order(conetDeg, decreasing=TRUE)] |> head(15)

conet <- induced_subgraph(conet, conetDeg>1)

wt <- igraph::walktrap.community(conet)
igraph::V(conet)$walktrap <- wt$membership
pal <- RColorBrewer::brewer.pal(length(unique(wt$membership)),
                                "Dark2") 
pal <- colorRampPalette(pal)(20)

ggraph(conet)+
    geom_edge_link(color="grey80")+
    geom_node_point(aes(color=factor(walktrap)), size=3)+
    geom_node_text(aes(label=name, color=factor(walktrap)),
                   check_overlap=TRUE, repel=TRUE,
                   bg.color = "white", segment.color="black",
                   bg.r = .15)+
    scale_color_manual(values=pal,
                         name="Walktrap")+
    theme_graph()
```

Dynamic layout can be also used to compare the networks, by `graphlayouts`, for comparing the multiple graphs, especially useful for time-series analysis. See the documentation of [`layout_as_dynamic`](http://graphlayouts.schochastics.net/reference/layout_dynamic.html) for specifying the alpha, which is default to 0.5. 

```{r dyn, warning=FALSE, fig.height=8, fig.width=15}
library(igraph)
dyn <- plotDynamic(list(abstnet, titlenet), concat="intersection",
                   titles=c("Abstract","Title"), alpha=0.8)
dyn
```


```{r}
sessionInfo()
```