---
title: "wcGeneSummary"
author: "Noriaki Sato"
package: wcGeneSummary
output: 
    BiocStyle::html_document:
        toc_float: true
vignette: >
    %\VignetteIndexEntry{wcGeneSummary}
    %\VignetteEngine{knitr::knitr}
    %\VignetteEncoding{UTF-8}  
---

This package aims to visualize the word and text information contained in the gene or the other omics identifiers such as microbiome, and identify important words among the clusters, integrate, and compare the clusters based on those information. It contributes to understanding the functional implications of omics identifier lists and aid in interpretation and visualization. In this vignette, the basic usage for generating the network and combining them are introduced, as well as customized usage. The detailed options and usage like the integration with the other packages are available in the package's [bookdown](https://noriakis.github.io/software/wcGeneSummary). The web server is available for the convenient querying [here](https://nsato.shinyapps.io/osplotweb).

# Basic usage

Install and load the package and the database for converting identifiers. In this example, we use mostly human-derived data, and use `org.Hs.eg.db`.

```{r load, warning=FALSE, message=FALSE}
# devtools::install_github("noriakis/wcGeneSummary")
library(wcGeneSummary)
library(org.Hs.eg.db)
library(ggplot2)
library(ggraph)
```

## Producing correlation networks

The main function generates a network. Various functions are available for this purpose, and to create a network from a gene list, wcGeneSummary is used. By specifying plotType as network, a network is generated. This generates a biotext class object with slots containing various information, which is then returned. The default ID type is SYMBOL, but it can be specified arbitrarily. The `net` slot stores the visualization result generated by `ggraph`.

```{r basic2, warning=FALSE, message=FALSE, cache=TRUE, fig.height=8, fig.width=8}
## Configure input genes
inpSymbol <- c("ERCC1","ERCC2","ERCC3","ERCC4","ERCC5","ERCC6","ERCC8")
net <- wcGeneSummary(inpSymbol, plotType="network")
net
```

The network visualization can be customized using the options like `colorText` or `edgeLink`.

```{r basic3, warning=FALSE, message=FALSE, cache=TRUE, fig.height=8, fig.width=8}
cxcls <- c()
for (i in c(1,2,3,5,6,8,9,10,11,12,13,14,16)){
    cxcls <- c(cxcls, paste0("CXCL",i))
}
cxclNet <- wcGeneSummary(cxcls, plotType="network",
    colorText=TRUE, edgeLink=FALSE)
cxclNet@net
```
The genes associated with the frequently occurred words within the cluster can be shown by `genePlot=TRUE`.

```{r basic5, warning=FALSE, message=FALSE, cache=TRUE, fig.height=8, fig.width=8}
net <- wcGeneSummary(cxcls, plotType="network",
    colorText=TRUE, edgeLink=FALSE, genePlot=TRUE)
net
net@net
```

## An example of a corelation network of words in the pathway

```{r basic6, warning=FALSE, message=FALSE, cache=TRUE, fig.height=8, fig.width=8}
keggPathways <- org.Hs.egPATH2EG
mappedKeys <- mappedkeys(keggPathways)
keggList <- as.list(keggPathways[mappedKeys])
## Hepatitis C
hCNet <- wcGeneSummary(keggList$`05160`, plotType="network",
                        layout="nicely", corThresh = 0.2, keyType="ENTREZID",
                        numWords = 30, excludeFreq = 5000, colorText=TRUE,
                        edgeLink=FALSE, showLegend=FALSE)
hCNetTrans <- hCNet@net + theme(plot.background = element_rect(fill = "transparent",colour = NA))
hCNetTrans
```
The summarization of text in enrichment analysis results can be performed by `enrich` option. The below example shows enrichment analysis of Reactome database.

```{r basic7}
hCNetReac <- wcGeneSummary(keggList$`05160`, enrich="reactome", keyType="ENTREZID",
                           topPath=30, numWords=30,
                           plotType="network", corThresh=0.2)
hCNetReacTrans <- hCNetReac@net +
                    theme(plot.background = element_rect(fill = "transparent",colour = NA))
hCNetReacTrans
```

Comparison of networks can be performed by `compareWordNet`.

```{r basic4, warning=FALSE, message=FALSE, cache=TRUE, fig.height=8, fig.width=8}
compareWordNet(list(hCNet, hCNetReac),
               titles=c("RefSeq","Reactome"),
               colPal = "Dark2",
               colNum=3)
```

# The other visualization options

## Annotating the gene cluster relationship

The customized functions are available, which annotate the gene cluster relationship. If you perform some clustering analysis for gene expression data and investigate the relationship between eigengenes, the `plotEigengeneNetworksWithWords` function can be used to populate the resulting dendrogram.

```{r custom1, warning=FALSE, message=FALSE, cache=TRUE, fig.height=8, fig.width=8}
mod <- returnExample()
plotEigengeneNetworksWithWords(mod$MEs, mod$colors)
```

## Producing word clouds

We use ERCC genes as input. The basic usage of the package is producing a word cloud of gene summaries by querying gene IDs, default to symbol. The other type can be set by `keyType`. As many of the words are commonly observed, you should limit word frequency by `excludeFreq`, which is default to 2000. TF-IDF on the all the summary is precomputed, and `exclude="tfidf"` can be specified too.

```{r basic_wc1, warning=FALSE, message=FALSE, cache=TRUE, fig.height=8, fig.width=8}
gwc <- wcGeneSummary(inpSymbol)
gwc
gwc@wc
```

The options in `wordcloud` including color and rotation can be specified in `argList`.

```{r basic_wc2, warning=FALSE, message=FALSE, cache=TRUE, fig.height=12, fig.width=12}
gwc <- wcGeneSummary(inpSymbol, numWords=200,
                     argList=list(max.words=200, random.order=FALSE,
                     colors=RColorBrewer::brewer.pal(5, "Dark2"),
                     rot.per=0.4))
gwc
gwc@wc
```



```{r}
sessionInfo()
```