---
title: "wcGeneSummary"
author: "Noriaki Sato"
package: wcGeneSummary
output: 
    BiocStyle::html_document:
        toc_float: true
vignette: >
    %\VignetteIndexEntry{wcGeneSummary}
    %\VignetteEngine{knitr::knitr}
    %\VignetteEncoding{UTF-8}  
---

This package aims to visualize the word and text information contained in the gene or the other omics identifiers such as microbiome, and identify important words among the clusters, and compare the clusters based on those information. It contributes to understanding the gene clusters and aid in easy interpretation and visualization. In this vignette, the two basic plots are introduced. The detailed usage like the network comparison is available in the package's [bookdown](https://noriakis.github.io/software/wcGeneSummary), and the web server is available for the easy querying.

# Basic usage

Load the package and the database for converting identifiers.
In this example, we use mostly human-derived data, and use `org.Hs.eg.db`.

```{r load, warning=FALSE, message=FALSE}
library(wcGeneSummary)
library(org.Hs.eg.db)
library(ggplot2)
library(ggraph)
```

## Producing word clouds

We use ERCC genes as input. The basic usage of the package is producing a word cloud of gene summaries by querying gene IDs, default to symbol. The other type can be set by `keyType`. As many of the words are commonly observed, you should limit word frequency by `excludeFreq`, which is default to 2000. TF-IDF on the all the summary is precomputed, and `exclude="tfidf"` can be specified too.

```{r basic, warning=FALSE, message=FALSE, cache=TRUE, fig.height=8, fig.width=8}
## Configure input genes
inpSymbol <- c("ERCC1","ERCC2","ERCC3","ERCC4","ERCC5","ERCC6","ERCC8")
gwc <- wcGeneSummary(inpSymbol)
gwc
gwc@wc
```

## Producing correlation networks

The next main function is producing correlation networks between words based on the co-occurrence of the words in the text.

```{r basic2, warning=FALSE, message=FALSE, cache=TRUE, fig.height=8, fig.width=8}
net <- wcGeneSummary(inpSymbol, plotType="network")
net
net@net
```

The network can be customized using the options like `colorText` or `edgeLink`.

```{r basic3, warning=FALSE, message=FALSE, cache=TRUE, fig.height=8, fig.width=8}
cxcls <- c()
for (i in c(1,2,3,5,6,8,9,10,11,12,13,14,16)){
    cxcls <- c(cxcls, paste0("CXCL",i))
}
cxclNet <- wcGeneSummary(cxcls, plotType="network",
    colorText=TRUE, edgeLink=FALSE)
cxclNet@net
```
Comparison of networks can be performed by `compareWordNet`.

```{r basic4, warning=FALSE, message=FALSE, cache=TRUE, fig.height=8, fig.width=8}
compareWordNet(list(cxclNet, net), titles=c("CXCL","ERCC"))
```

The genes associated with the frequently occurred words within the cluster can be shown by `genePlot=TRUE`.

```{r basic5, warning=FALSE, message=FALSE, cache=TRUE, fig.height=8, fig.width=8}
net <- wcGeneSummary(cxcls, plotType="network",
    colorText=TRUE, edgeLink=FALSE, genePlot=TRUE)
net
net@net
```

## An example of a corelation network of words in the pathway

```{r basic6, warning=FALSE, message=FALSE, cache=TRUE, fig.height=8, fig.width=8}
library(org.Hs.eg.db)
keggPathways <- org.Hs.egPATH2EG
mappedKeys <- mappedkeys(keggPathways)
keggList <- as.list(keggPathways[mappedKeys])
## Hepatitis C
hCNet <- wcGeneSummary(keggList$`05160`, plotType="network",
                        layout="nicely", corThresh = 0.2, keyType="ENTREZID",
                        numWords = 30, excludeFreq = 5000, colorText=TRUE,
                        edgeLink=FALSE, showLegend=FALSE)
hCNetTrans <- hCNet@net + theme(plot.background = element_rect(fill = "transparent",colour = NA))
hCNetTrans
```
The summarization of text in enrichment analysis results can be performed by `enrich` option. The below example shows enrichment analysis of Reactome database.

```{r basic7}
hCNetReac <- wcGeneSummary(keggList$`05160`, enrich="reactome", keyType="ENTREZID",
                           topPath=30, numWords=30,
                           plotType="network", corThresh=0.2)
hCNetReacTrans <- hCNetReac@net +
                    theme(plot.background = element_rect(fill = "transparent",colour = NA))
```

# Annotating the gene cluster relationship

The customized functions are available, which annotate the gene cluster relationship. If you perform some clustering analysis for gene expression data and investigate the relationship between eigengenes, the `plotEigengeneNetworksWithWords` function can be used to populate the resulting dendrogram.

```{r custom1, warning=FALSE, message=FALSE, cache=TRUE, fig.height=8, fig.width=8}
mod <- returnExample()
plotEigengeneNetworksWithWords(mod$MEs, mod$colors)
```

```{r}
sessionInfo()
```