---
title: "Example workflow investigating BK polyomavirus infection"
author: "Noriaki Sato"
package: wcGeneSummary
output: 
    BiocStyle::html_document:
        toc_float: true
vignette: >
    %\VignetteIndexEntry{wcGeneSummary}
    %\VignetteEngine{knitr::knitr}
    %\VignetteEncoding{UTF-8}  
---

This vignette shows the example workflow when investigating transcriptomic dataset related to BK polyomavirus (BKPyV) infection in renal proximal tubular epithelial cells ([An et al. 2019](https://pubmed.ncbi.nlm.nih.gov/30620752/)).
Differentially expressed genes at 3 days post-infection are identified.


# Load the necessary packages and genes

```{r load, warning=FALSE, message=FALSE}
library(wcGeneSummary)
library(org.Hs.eg.db)
library(ggplot2)
library(ReactomePA);library(clusterProfiler)
library(ggraph)
degs <- wcGeneSummary:::d3degAn2019
length(degs)
degs
```

## Enrichment analysis

First, we perform enrichment analysis using ReactomePA.
From the enrichment analysis results, the cluster is related to transcriptional regulation by TP53.

```{r ea, warning=FALSE, message=FALSE, cache=TRUE}
## Convert to ENTREZID
entre <- AnnotationDbi::select(org.Hs.eg.db, keytype="SYMBOL",
                               keys = degs, columns = "ENTREZID")$ENTREZID
pway <- setReadable(enrichPathway(entre), org.Hs.eg.db)
sigpway <- subset(pway@result, p.adjust<0.05)
sigpway$Description
cnetplot(pway)

## Genes involved in significant pathways
unlist(unique(sapply(sigpway$geneID, function (x) strsplit(x,"/"))))
```
We store the name of enriched pathways in the network for the downstream analysis.

```{r netenrich}
netreac <- wcGeneSummary(degs, enrich="reactome", plotType="network", numWords=50,
                         colorText=TRUE)
```

Next we perform the plain function producing a correlation network, with showing the genes related to high-frequency words in the text.

```{r basic2, warning=FALSE, message=FALSE, cache=TRUE}
net1 <- wcGeneSummary(degs, plotType="network",
                      colorText=TRUE, numWords=20,
                      genePlot=TRUE, genePlotNum=5)
net1$net
top5 <- names(head(net1$geneCount, n=5))
top5
```

These genes are further queried for PubMed information.

```{r table}
abst <- wcAbst(top5)
```

Show the network for the article titles.

```{r absttitle}
titlenet <- wcAbst(redo=abst, target="title", tfidf=TRUE,
       plotType="network", colorText=TRUE, corThresh=0.1,
       preset=TRUE)
titlenet$net
```

Show the network for the article abstract.

```{r abstmain}
abstnet <- wcAbst(redo=abst, target="abstract", tfidf=TRUE,
       plotType="network", colorText=TRUE, corThresh=0.1,
       preset=TRUE)
abstnet$net
```

From the RefSeq summary and articles related to important genes, the cluster could have functionality of DNA damage response, which is also upregulated by BKPyV infection. These networks can be combined to find intersections and differences. We can see that in addition to Reactome pathway names, plenty of information could be obtained and summarized by querying other databases, which could aid in interpreting clusters of genes and hypothesis generation.

```{r combine}
compareWordNet(list(abstnet, titlenet, netreac, net1),
               titles=c("Abstract","Title","Reactome","RefSeq"))
```


```{r dyn, warning=FALSE, fig.height=6, fig.width=6}
library(igraph)
dyn <- plotDynamic(list(abstnet, titlenet, net1),
                   titles=c("Abstract","Title","RefSeq"))
dyn
```


```{r}
sessionInfo()
```