---
title: "Pathview text function"
author: "Noriaki Sato"
package: wcGeneSummary
output: 
    BiocStyle::html_document:
    toc_float: true
vignette: >
    %\VignetteIndexEntry{wcGeneSummary}
    %\VignetteEngine{knitr::knitr}
    %\VignetteEncoding{UTF-8}  
---

KEGG mapping plots produced by pathview can be populated with text information ([Luo and Brouwer. 2013](https://academic.oup.com/bioinformatics/article/29/14/1830/232698])).

```{r load, warning=FALSE, message=FALSE}
library(wcGeneSummary)
library(pathview)
```

Suppose an example for KEGG Orthology obtained from some microbiome-related experiments.
By using [MicrobiomeProfiler](https://github.com/YuLab-SMU/MicrobiomeProfiler), enriched pathway using over-representation analysis can be obtained as follows.

```{r micro, warning=FALSE, message=FALSE}
kos <- c("K00640","K23304","K01760","K01697","K00549","K00899")
enriched <- MicrobiomeProfiler::enrichKO(kos)
clusterProfiler::cnetplot(enriched)
```
Now plot these KOs in the pathview.

```{r pathview, warning=FALSE, message=FALSE}
## As the output is PNG, this script is omitted.
pvInput <- rep(1,length(kos))
names(pvInput) <- kos
# pathview(kos, pathway.id = "00270", species = "ko")
```
![raw pathview plot](https://github.com/noriakis/software/blob/main/images/ko00270.pathview.png?raw=true)

To use the `pathviewText()` one should specify which database to retrieve the text, `"refseq"` or `"abstract"`.
The default is "abstract", which fetches information from RefSeq database. In this case, KO number is not included in RefSeq, thus the abstract is chosen. However, querying in KO number itself is not useful. We conver the KO number to EC number using KEGG REST API, and fetches the description using `wcEC()` function, and query the enzyme name in PubMed. In this way, one must provide `searchTerm`, and `termMap`, which map the KO and EC. Also, `node.types="ortholog"` must be passed to `pathview`.

```{r pathview2, warning=FALSE, message=FALSE}
## Again, the codes are commented out.

# ecko <- data.table::fread("https://rest.kegg.jp/link/ko/ec",
#                             header=FALSE)
# ecko$EC <- sapply(strsplit(ecko$V1, ":"),"[",2)
# ecko$KO <- sapply(strsplit(ecko$V2, ":"), "[", 2)
# filt <- ecko[ ecko$KO %in% kos, ]
# ecnum <- filt$EC
# termMap <- filt[,c("KO","EC")] |> `colnames<-`(c("query","number"))
# ecQuery <- wcEC("enzyme.dat", ecnum = ecnum, onlyTerm =TRUE)
# ecMap <- wcEC("enzyme.dat", ecnum = ecnum, onlyDf =TRUE)
# termMap <- merge(termMap, ecMap[,c("number","desc")], by="number") |>
#   `colnames<-`(c("number","query","description"))

## Main function

# pathviewText(kos, keyType = "KO",
#      target="abstract",
#      pid = "00270",
#      org = "ko", 
#      searchTerms = ecQuery,
#      termMap = termMap,
#      node.types="ortholog")

## We can transpose the barplot by trans=TRUE

# pathviewText(kos, keyType = "KO",
#      target="abstract",
#      pid = "00270",
#      org = "ko", 
#      searchTerms = ecQuery,
#      termMap = termMap,
#      node.types="ortholog")
```

The below picture shows transposed results.

![annotated pathview plot - KO](https://github.com/noriakis/software/blob/main/images/pathviewText.png?raw=true)

For gene queries, we can obtain the same plot using relatively simple query.

```{r gene, warning=FALSE, message=FALSE}
query <- c("TP53","CDC45","CDC6")
```